src/base/undistortion.cc

```diff
@@ -634,22 +634,75 @@ Camera UndistortCamera(const UndistortCameraOptions& options,
   CHECK_GT(options.min_scale, 0.0);
   CHECK_LE(options.min_scale, options.max_scale);
   CHECK_NE(options.max_image_size, 0);
+  CHECK_LT(options.max_fov, 180.0);
+  CHECK_GT(options.max_fov, 0.0);
 
   Camera undistorted_camera;
   undistorted_camera.SetModelId(PinholeCameraModel::model_id);
   undistorted_camera.SetWidth(camera.Width());
   undistorted_camera.SetHeight(camera.Height());
 
-  // Copy focal length parameters.
-  const std::vector<size_t>& focal_length_idxs = camera.FocalLengthIdxs();
-  CHECK_LE(focal_length_idxs.size(), 2)
-      << "Not more than two focal length parameters supported.";
-  if (focal_length_idxs.size() == 1) {
-    undistorted_camera.SetFocalLengthX(camera.FocalLength());
-    undistorted_camera.SetFocalLengthY(camera.FocalLength());
-  } else if (focal_length_idxs.size() == 2) {
-    undistorted_camera.SetFocalLengthX(camera.FocalLengthX());
-    undistorted_camera.SetFocalLengthY(camera.FocalLengthY());
+  // Copy focal length parameters, or estimate them
+  double max_valid_radius = std::numeric_limits<double>::max();
+  Eigen::Vector2d principal_point = Eigen::Vector2d::Zero();
+  if (options.estimate_fov) {
+    // Estimate maximal valid diagonal FOV (based on radius~angle monothonicity
+    const std::vector<size_t>& principal_idxs = camera.PrincipalPointIdxs();
+    CHECK_EQ(principal_idxs.size(), 2)
+        << "Only two principal point parameters supported.";
+    const double max_fov = options.max_fov * M_PI / 180.0;
```

Comment:
There is a function called `DegToRad` in ``util/math.h``.

Reply:
Yes, my solution looks a bit confusing:

* `max_fov` limits diagonal field-of-view by a user-specified constant
* `clip_by_vfov` [additionally] limits diagonal field-of-view by current horizontal field-of-view
* `clip_by_hfov` [additionally] limits diagonal field-of-view by current vertical field-of-view

I'll think how to make it more clear and update pull request